learning/tour-of-beam/backend/internal/service/content.go

```diff
@@ -23,30 +23,48 @@ import (
 	"beam.apache.org/learning/tour-of-beam/backend/internal/storage"
 )
 
-var ErrNoUnit = errors.New("unit not found")
-
 type IContent interface {
-	GetContentTree(ctx context.Context, sdk tob.Sdk, userId *string) (tob.ContentTree, error)
-	GetUnitContent(ctx context.Context, sdk tob.Sdk, unitId string, userId *string) (tob.Unit, error)
+	GetContentTree(ctx context.Context, sdk tob.Sdk) (tob.ContentTree, error)
+	GetUnitContent(ctx context.Context, sdk tob.Sdk, unitId string) (tob.Unit, error)
+	GetUserProgress(ctx context.Context, sdk tob.Sdk, userId string) (tob.SdkProgress, error)
+	SetUnitComplete(ctx context.Context, sdk tob.Sdk, unitId, uid string) error
 }
 
 type Svc struct {
 	Repo storage.Iface
 }
 
-func (s *Svc) GetContentTree(ctx context.Context, sdk tob.Sdk, userId *string) (ct tob.ContentTree, err error) {
-	// TODO enrich tree with user-specific state (isCompleted)
+func (s *Svc) GetContentTree(ctx context.Context, sdk tob.Sdk) (ct tob.ContentTree, err error) {
 	return s.Repo.GetContentTree(ctx, sdk)
 }
 
-func (s *Svc) GetUnitContent(ctx context.Context, sdk tob.Sdk, unitId string, userId *string) (tob.Unit, error) {
-	// TODO enrich unit with user-specific state: isCompleted, userSnippetId
+func (s *Svc) GetUnitContent(ctx context.Context, sdk tob.Sdk, unitId string) (tob.Unit, error) {
 	unit, err := s.Repo.GetUnitContent(ctx, sdk, unitId)
 	if err != nil {
 		return tob.Unit{}, err
 	}
 	if unit == nil {
-		return tob.Unit{}, ErrNoUnit
+		return tob.Unit{}, tob.ErrNoUnit
 	}
 	return *unit, nil
 }
+
+func (s *Svc) GetUserProgress(ctx context.Context, sdk tob.Sdk, userId string) (tob.SdkProgress, error) {
+	progress, err := s.Repo.GetUserProgress(ctx, sdk, userId)
+	if errors.Is(err, tob.ErrNoUser) {
+		// make an empty list a default response
+		return tob.SdkProgress{Units: make([]tob.UnitProgress, 0)}, nil
+	}
+	if err != nil {
+		return tob.SdkProgress{}, err
+	}
+	if progress == nil {
+		panic("progress is nil, no err")
```

Comment:
Do we really want to panic here, or should we just return this string as an error `tob.SdkProgress{}, <error>` (and log the error)? Panicing will bring everything down here when this could be a bug for this specific user, right?

Reply:
The case when `progress == nil && err == nil` is definetely a logical problem in an app. We need to be notified if this happens. Panic is reported in Error reporting console and also triggers cold restart of an app.

[error reporting howto](https://cloud.google.com/functions/docs/monitoring/error-reporting)

Reply:
I guess that's fair - this would represent a pretty complete failure. I'm ok with it